packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/AppImpl.kt

```diff
@@ -40,11 +40,12 @@ internal class AppImpl(
         )
 
     override suspend fun login(credentials: Credentials): Result<User> {
+        val credentialsInternal: CredentialImpl = io.realm.internal.util.Validation.checkType(credentials, "credentials")
         return RealmInterop.runCatching {
             suspendCoroutine { continuation ->
                 realm_app_log_in_with_credentials(
                     nativePointer,
-                    credentials.nativePointer,
+                    (credentials as CredentialImpl).nativePointer,
```

Comment:
I see your point, but lets discuss. I guess it would be nice to clarify the classes that would make sense to be able to mock and then be very explicit about that the rest of the interfaces are not ment for user implementation/extension. 

Maybe we can put together a generic error message signaling this in `Validation`.

Reply:
Discussed offline. 

This is fine for now. There should be virtually zero reasons for people to pass in their own custom implementations, and it creates a nice separation in our codebase, where it is very easy to see when the public API is affected.